Validate and log when an unrecognized parameter/flag is used

wugalde19 opened the pull request:
Resolves https://github.com/NethermindEth/nethermind/issues/6942

## Changes
Changes made into `src/Nethermind/Nethermind.Runner/Program.cs`

- Add a function to load all possible flags/params/arguments to be used and put them on a hash
  - The hash is initialized with the default flags allowed in the app and not related to flags in config files
- Add a function to validate current args against the hash generated
- Call the validation function into `Run` so that we initialize the app and validate before start running most of the logic.

## Types of changes

#### What types of changes does your code introduce?

- [ ] Bugfix (a non-breaking change that fixes an issue)
- [X] New feature (a non-breaking change that adds functionality)
- [ ] Breaking change (a change that causes existing functionality not to work as expected)
- [ ] Optimization
- [ ] Refactoring
- [ ] Documentation update
- [ ] Build-related changes
- [ ] Other: _Description_

## Testing

#### Requires testing

- [X] Yes
- [ ] No

#### If yes, did you write tests?

- [ ] Yes
- [X] No

#### Notes on testing

I did not see any test file for Program.cs files

## Documentation

#### Requires documentation update

- [ ] Yes
- [X] No

#### Requires explanation in Release Notes

- [ ] Yes
- [ ] No

## Remarks

I recorded a video showing how it works and the error log is displayed when the argument/option/flag used is not valid and when it is valid.
Also, the cases where no flag is provided and when a default app flag (like `--help`) is provided, and when we try using duplicate arguments

https://www.loom.com/share/0ac1011bac1247b992a9fbadba9e5b26?sid=eb48d13b-961e-490e-a1ff-5fe51a2580b0

List of commands used:
```
# Should work
./publish/nethermind --help

# Should work
./publish/nethermind -v

# Should work
./publish/nethermind

# Should work
./publish/nethermind --JsonRpc.WebSocketsPort 8546

# Duplicate argument error
./publish/nethermind --JsonRpc.WebSocketsPort 8546 --JsonRpc.WebSocketsPort 8546

# Unrecognized argument error
./publish/nethermind --JsonRpc.WebsocketsPort 8546
```

LukaszRozmej commented on src/Nethermind/Nethermind.Runner/Program.cs:
```diff
@@ -591,4 +593,57 @@ private static string GetProductInfo()
+        return options;
+    }
+
+    private static void ValidateCommandLineFlags(string[] args)
```
If we are validating here I would also move `GetDuplicateArguments` and duplication validation under this one method.

LukaszRozmej commented on src/Nethermind/Nethermind.Runner/Program.cs:
```diff
@@ -591,4 +593,57 @@ private static string GetProductInfo()
+        foreach (var arg in args)
+        {
+            // Check if the argument is a parameter (starts with --)
+            if (arg.StartsWith("--") && !validParameters.Contains(arg))
```
In `GetDuplicateArguments` we actually already have local method `GetArgumentNames` that should be reused here.

wugalde19 replied:
@LukaszRozmej I used it and it seems to work just fine 👍🏻

LukaszRozmej commented on src/Nethermind/Nethermind.Runner/Program.cs:
```diff
@@ -591,4 +593,57 @@ private static string GetProductInfo()
         return info.ToString();
     }
+
+    private static HashSet<string> generateValidConfigOptionsHash()
```
We use PascalCase in C# for method names.

LukaszRozmej commented on src/Nethermind/Nethermind.Runner/Program.cs:
```diff
@@ -591,4 +593,57 @@ private static string GetProductInfo()
+
+    private static HashSet<string> generateValidConfigOptionsHash()
+    {
+        var options = new HashSet<string>();
+
+        Type configurationType = typeof(IConfig);
+        IEnumerable<Type> configTypes = TypeDiscovery.FindNethermindBasedTypes(configurationType)
+            .Where(ct => ct.IsInterface);
+
+        foreach (Type configType in configTypes.Where(ct => !ct.IsAssignableTo(typeof(INoCategoryConfig))).OrderBy(c => c.Name))
+        {
+            if (configType is null)
+            {
+                continue;
+            }
+
+            ConfigCategoryAttribute? typeLevel = configType.GetCustomAttribute<ConfigCategoryAttribute>();
+
+            if (typeLevel is not null && (typeLevel?.DisabledForCli ?? true))
+            {
+                continue;
+            }
+
+            foreach (PropertyInfo propertyInfo in configType
+                            .GetProperties(BindingFlags.Public | BindingFlags.Instance)
+                            .OrderBy(p => p.Name))
+            {
+                ConfigItemAttribute? configItemAttribute = propertyInfo.GetCustomAttribute<ConfigItemAttribute>();
+                if (!(configItemAttribute?.DisabledForCli ?? false))
+                {
+                    options.Add($"--{configType.Name[1..].Replace("Config", string.Empty)}.{propertyInfo.Name}");
+                }
+            }
+        }
+
+        return options;
+    }
```
Wouldn't it be better to use `CommandLineApplication.GetOptions()` to get all options and check on those instead of parsing them again here from config? And not all options are from configs, some are added explicitly!

wugalde19 replied:
Hey @LukaszRozmej 
I just run a diff against what I had pushed and seems like I forgot to push the most updated version of the latest commit
I did actually noticed that some options are added explicitly and for that reason I added this
<img width="462" alt="image" src="https://github.com/user-attachments/assets/44a85e3e-e3ad-43c2-b22a-148685430099">

But not that you mention the `.GetOptions()` methods seems like a better way to get them
Gonna give it a try 
Thanks

wugalde19 replied:
You were right @LukaszRozmej 
I didn't need a new method to get all the options, the way you proposed did it for me
Here's a sample of the options I printed in the console to test
<img width="804" alt="image" src="https://github.com/user-attachments/assets/216f40a7-4c09-47f8-81ab-c4063bbed0e3">

LukaszRozmej replied:
Yep, didn't know it was an option until I looked for it either. But it makes sense it will be there.

wugalde19 commented:
@LukaszRozmej I've already addressed all your comments and updated the video to validate new changes
Feel free to review it again and let me know if you have more feedback
Thanks

LukaszRozmej commented on src/Nethermind/Nethermind.Runner/Program.cs:
```diff
@@ -591,4 +586,32 @@ private static string GetProductInfo()
+        // Get all valid options from the configuration files
+        var validArguments = app.GetOptions().SelectMany(o => new[] { o.LongName, o.ShortName }).ToArray();
+
+        var argumentsNamesProvided = GetArgumentNames(args).Select(a => a.ToString()).ToArray();
```
You don't need to allocate strings, `MemoryContentsComparer` can compare `Memory<char>` directly without additional allocations

LukaszRozmej replied:
Avoid `ToArray`, either keep it `IEnumerable` or use `ToPooledList` and use pooled memory (need to be disposed with `using` statement).

LukaszRozmej commented on src/Nethermind/Nethermind.Runner/Program.cs:
```diff
@@ -591,4 +586,32 @@ private static string GetProductInfo()
+        foreach (var argumentName in argumentsNamesProvided)
+        {
+            // Check if the argument provided is a valid option/argument
+            if (!validArguments.Contains(argumentName))
```
`validArguments` should be `HashSet` for quick lookup. Right now your code is O(N^2), with `HashSet` it will be O(N)

LukaszRozmej commented on src/Nethermind/Nethermind.Runner/Program.cs:
```diff
@@ -591,4 +586,32 @@ private static string GetProductInfo()
+            }
+        }
+
+        string duplicateArgumentsList = string.Join(", ", GetDuplicateArguments(args));
```
Don't pass args here again, pass `argumentsNamesProvided`

wugalde19 replied:
Good call
No need to process/get the names again if we already did it

LukaszRozmej commented on src/Nethermind/Nethermind.Runner/Program.cs:
```diff
@@ -591,4 +586,32 @@ private static string GetProductInfo()
+    private static bool ValidateArguments(string[] args, CommandLineApplication app)
+    {
+        // Get all valid options from the configuration files
+        var validArguments = app.GetOptions().SelectMany(o => new[] { o.LongName, o.ShortName }).ToArray();
```
avoid creating intermittent arrays of `new[] { o.LongName, o.ShortName }` with a simple loop.

LukaszRozmej commented on src/Nethermind/Nethermind.Runner/Program.cs:
```diff
@@ -256,29 +251,29 @@ private static void Run(string[] args)
     }
 
-    private static IEnumerable<ReadOnlyMemory<char>> GetDuplicateArguments(string[] args)
+    private static ReadOnlyMemory<char> GetArgumentName(string arg) => arg.StartsWith("--") ? arg.AsMemory(2) : arg.StartsWith('-') ? arg.AsMemory(1) : ReadOnlyMemory<char>.Empty;
+    private static IEnumerable<ReadOnlyMemory<char>> GetArgumentNames(IEnumerable<string> args)
```
Make those methods local in `ValidateArguments`, they don't make sense outside of it.

wugalde19 replied:
My main concern when I extracted those was `GetDuplicateArguments` not being able to reach that code.
I should have tested it before.
Thanks
Done

LukaszRozmej reviewed (changes requested):
Better design, but way to much allocations.

wugalde19 commented:
@LukaszRozmej Thanks for all the feedback provided
Definitely learning a lot and it is forcing me to be more thoughtful with the code
Ready for another round of reviews

LukaszRozmej commented on src/Nethermind/Nethermind.Runner/Program.cs:
```diff
@@ -160,6 +152,9 @@ private static void Run(string[] args)
 
         BuildOptionsFromConfigFiles(app);
 
+        bool validArgs = ValidateArguments(args, app);
+        if (!validArgs) return;
```
minor: would drop `validArgs` variable it doesn't add anything.
```suggestion
        if (!ValidateArguments(args, app)) return;
```

LukaszRozmej commented on src/Nethermind/Nethermind.Runner/Program.cs:
```diff
@@ -591,4 +565,69 @@ private static string GetProductInfo()
+        }
+
+        // Get all valid options from the configuration files
+        var validArguments = GetValidArguments(app);
+
+        var argumentsNamesProvided = GetArgumentNames(args).ToList();
```
Try not to use `var` unless the types are anonymous or very long and obvious, makes harder to read and understand what is going on in GitHub

wugalde19 replied:
Not quite sure I understand
Do you mean avoid using var and try to set an explicit type when possible?

The last part `makes harder to read and understand what is going on in GitHub` it's the part that confuses me

stdevMac replied:
@wugalde19 on Github, you can't infer the type or jump into the function definition to see what it is. 

It makes sense to user var if it is evident, like GetString(), that it should return some string. 

For the example in question, what does return `GetValidArguments`? The context of the changes does not make it clear.  

Makes more sense? I think it's what @LukaszRozmej was referring to.

wugalde19 replied:
Ohhhh I see
Of course, the idea is to provide more context of what a variable will contain unless it's easy to infer.
Yeah yeah makes sense
And that part about Github it's like that you cannot jump on Github between definitions easily as you can do it in some IDEs which makes things harder in terms of knowing/checking what a method returns and all of that

Perfect
Thanks for clarifying this to me @stdevMac

LukaszRozmej commented on src/Nethermind/Nethermind.Runner/Program.cs:
```diff
@@ -591,4 +565,69 @@ private static string GetProductInfo()
+        foreach (var argumentName in argumentsNamesProvided)
+        {
+            // Check if the argument provided is a valid option/argument
+            if (!validArguments.Contains(argumentName.ToString()))
```
Ah we need that `ToString` conversion here...
But maybe we can made `HashSet<ReadOnlyMemory<char>>` to avoid it?

wugalde19 replied:
I did try but it didn't work the way I did it yesterday, however, today I realized that I had to tweak the implementation of `GetValidArguments` a little bit to make it work, and it works now

LukaszRozmej commented on src/Nethermind/Nethermind.Runner/Program.cs:
```diff
@@ -591,4 +565,69 @@ private static string GetProductInfo()
+
+    private static bool ValidateArguments(string[] args, CommandLineApplication app)
+    {
+        static HashSet<string> GetValidArguments(CommandLineApplication app)
+        {
+            var validArguments = new HashSet<string>();
+            foreach (var option in app.GetOptions())
+            {
+                if (!string.IsNullOrEmpty(option.LongName))
+                {
+                    validArguments.Add(option.LongName);
+                }
+                if (!string.IsNullOrEmpty(option.ShortName))
+                {
+                    validArguments.Add(option.ShortName);
+                }
+            }
+
+            return validArguments;
+        }
```
```suggestion
        static HashSet<ReadOnlyMemory<char>> GetValidArguments(CommandLineApplication app)
        {
            HashSet<ReadOnlyMemory<char>> validArguments = new(new MemoryContentsComparer<char>());
            foreach (var option in app.GetOptions())
            {
                if (!string.IsNullOrEmpty(option.LongName))
                {
                    validArguments.Add(option.LongName.AsMemory());
                }
                if (!string.IsNullOrEmpty(option.ShortName))
                {
                    validArguments.Add(option.ShortName.AsMemory());
                }
            }

            return validArguments;
        }
```

wugalde19 replied:
Interesing that solving previous comment I did apply those same changes you are suggesting
Except they way you are initializing the hash

Your suggestion
```
HashSet<ReadOnlyMemory<char>> validArguments = new(new MemoryContentsComparer<char>());
```

My code 
```
HashSet<ReadOnlyMemory<char>> validArguments = new HashSet<ReadOnlyMemory<char>>(new MemoryContentsComparer<char>());
```


I changed it to use your way to initialize it (simpler)

LukaszRozmej commented on src/Nethermind/Nethermind.Runner/Program.cs:
```diff
@@ -591,4 +565,69 @@ private static string GetProductInfo()
+        // Get all valid options from the configuration files
+        var validArguments = GetValidArguments(app);
+
+        var argumentsNamesProvided = GetArgumentNames(args).ToList();
```
I would either:
1. Leave it as `IEnumerable` and drop `ToList` - it can complain about re-enumeerating it, but as it is all in-memory there is no big downside
2. As it is local variable, use pooled memory - to avoid creating temp object:
```suggestion
        using ArrayPoolList<ReadOnlyMemory<char>> argumentsNamesProvided = GetArgumentNames(args).ToPooledList();
```

wugalde19 replied:
When I drop the `ToList` and the reason why I'm using it is to avoid this error
<img width="570" alt="image" src="https://github.com/user-attachments/assets/10a2dd82-aaad-4b3e-ba53-30d00f857892">

wugalde19 replied:
Using option 2 `using ArrayPoolList<ReadOnlyMemory<char>> argumentsNamesProvided = GetArgumentNames(args).ToPooledList();` gives me this other error
<img width="570" alt="image" src="https://github.com/user-attachments/assets/43d5c8fd-7507-4beb-b192-01a771ec642d">

wugalde19 replied:
I decided to go with option 1 and replaced that line with the following
```
IEnumerable<ReadOnlyMemory<char>> argumentsNamesProvided = GetArgumentNames(args);
```

LukaszRozmej replied:
`ToPooledList` is an extension method in `Nethermind.Core.Extensions.EnumerableExtensions`

LukaszRozmej reviewed (commented):
Cool, IMO code now is quite good, would apply few final touches to make it perfect ;)

wugalde19 commented on src/Nethermind/Nethermind.Runner/Program.cs:
```diff
@@ -256,30 +251,9 @@ private static void Run(string[] args)
     }
 
-    private static IEnumerable<ReadOnlyMemory<char>> GetDuplicateArguments(string[] args)
+    private static IEnumerable<ReadOnlyMemory<char>> GetDuplicateArguments(IEnumerable<ReadOnlyMemory<char>> argumentsNames)
```
You're right.
Done

wugalde19 commented:
@LukaszRozmej one step closer to a better implementation
You are great
Thanks for your patience

LukaszRozmej commented:
Congrats for persistence ;)

LukaszRozmej commented on src/Nethermind/Nethermind.Runner/Program.cs:
```diff
@@ -591,4 +557,76 @@ private static string GetProductInfo()
+            // Check if the argument provided is a valid option/argument
+            if (!validArguments.Contains(argumentName))
+            {
+                throw new ArgumentException($"Unrecognized argument: {argumentName}.\nRun --help for a list of available options and commands.");
```
One more thing, maybe better than throwing here would be doing something similar as in duplicate arguments:
1. logging an error.
2. setting ExitCode to specialized (new) exit code
3. returning false.

wugalde19 replied:
Actually, that's a great suggestion and makes a ton of sense
Let me add that real quick

wugalde19 commented:
@LukaszRozmej Just applied changes suggested on last two comments

LukaszRozmej commented:
@wugalde19 thank you for your contribution 🫡

benaadams commented:
Fails for?
```
 --JsonRpc.JwtSecretFile=C:/ethereum/jwt.hex
```

🤔

wugalde19 commented:
My pleasure @LukaszRozmej

benaadams commented:
Failed due to unrecognized argument - [JsonRpc.JwtSecretFile=C:/ethereum/jwt.hex].
Run --help for a list of available options and commands.

wugalde19 commented:
> JsonRpc.JwtSecretFile=C:/ethereum/jwt.hex

@benaadams Please try again
@LukaszRozmej came out with a much better solution ([here](https://github.com/NethermindEth/nethermind/commit/ebdf27ac23faa6381f37c75f3b716c85e8e1c26b))
Those changes seemed to have fixed the issue you were having
